code/cdw.py

```diff
@@ -0,0 +1,259 @@
+    good_inds = labels.query('labels not in @bad_labels').index
+    labels_subset = labels.loc[good_inds]
+    
+    print('Dropping unwanted frames...')
+    frames_subset = bold.get_fdata()[:,:,:,good_inds]
+    bold_subset = nib.Nifti1Image(frames_subset, bold.affine, bold.header)
+    
+    print('Done.')
+    return (bold_subset, labels_subset)
+    
+
+def reorder_by_labels(bold, labels):
+    """
+    Reorder labels and frames of 4D BOLD image
+    
+    Sorts labels (of frames) alphabetically and then reorders BOLD frames
+    to match that order.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original 4D BOLD image.
+    labels : DataFrame
+        Original frame labels.
+
+    Returns
+    -------
+    bold_sorted : Nifti1Image
+        Sorted 4D BOLD image..
+    labels_sorted : DataFrame
+        Sorted frame labels.
+
+    """
+    
+    labels['fmri_i'] = range(len(labels))
+    labels_sorted = labels.sort_values(by=['labels','chunks'])
+    labels_sorted.reset_index(inplace=True)
+    
+    print('Reordering data...')
+    frames_sorted = bold.get_fdata()[:,:,:,labels_sorted['fmri_i']]
+    bold_sorted = nib.Nifti1Image(frames_sorted, bold.affine, bold.header)
+    
+    print('Done.')
+    return (bold_sorted, labels_sorted)
+
+
+def create_model_rdm(labels):
+    """
+    Create RDM from labels
+    
+    Creates a theoretical representational dissimilarity matrix from given
+    labels: dissimilarity within label is 0, between labels 1.
+
+    Parameters
+    ----------
+    labels : DataFrame
+        Labels of frames (sorted).
+
+    Returns
+    -------
+    rdm_model : ndarray
+        Theoretical RDM.
+
+    """
+    
+    # To get an "identity matrix" over labels we first expand labels to dummies
+    # (one column per label with values=[0,1]) and then get the dot product of
+    # that and its transpose -- result is 1 when the labels of the frame match,
+    # and 0 when they don't. To get RDM, switch those.
+    dummies = pd.get_dummies(labels.labels).to_numpy()
+    rdm_model = np.dot(dummies, dummies.T)
+    rdm_model = abs(rdm_model-1)
+    
+    return rdm_model
+
+
+def get_patch_data(bold, mask, rad):
+    
+    # Get (x,y,z) inds of all voxels inside mask
+    inds = np.where(mask.get_fdata() == 1)
+    print(f'Number of voxels in mask:: {len(inds[0])}')
+    
+    # Create searchlight kernel
+    kernel = _make_searchlight(rad)
+    print(f'Number of voxels in searchlight: {len(kernel)}')
+    
+    # For each voxel, get and return patch data
+    for center_ind in zip(inds[0], inds[1], inds[2]):
+        
+        # Get inds for this patch
+        patch_inds = []
+        for (kx,ky,kz) in kernel:
+            patch_inds.append( (center_ind[0]+kx,
+                                center_ind[1]+ky,
+                                center_ind[2]+kz) )
+        
+        # Get the bold data for the patch
+        patch_data = np.vstack([bold.get_fdata()[x,y,z,:] for (x,y,z) in patch_inds]).T
```

Comment:
At this point, you start reading up about [NumPy advanced indexing](https://numpy.org/doc/stable/reference/arrays.indexing.html#advanced-indexing).
```suggestion
        patch_data = bold.get_fdata()[tuple(patch_inds.T)].T
```